src/copilot/inspect/InspectionCache.ts

```diff
@@ -0,0 +1,90 @@
+import { SymbolKind, TextDocument } from 'vscode';
+import { METHOD_KINDS, getClassesAndMethodsOfDocument, logger } from '../utils';
+import { Inspection } from './Inspection';
+import * as crypto from "crypto";
+import { SymbolNode } from './SymbolNode';
+
+/**
+ * A map based cache for inspections of a document.
+ * format: `Map<documentKey, Map<symbolQualifiedName, [symbolVersionId, Inspection[]]`
+ */
+const DOC_SYMBOL_VERSION_INSPECTIONS: Map<string, Map<string, [string, Inspection[]]>> = new Map();
+
+export default class InspectionCache {
+    public static hasCache(document: TextDocument, symbol?: SymbolNode): boolean {
+        const documentKey = document.uri.fsPath;
+        if (!symbol) {
+            return DOC_SYMBOL_VERSION_INSPECTIONS.has(documentKey);
+        }
+        const symbolInspections = DOC_SYMBOL_VERSION_INSPECTIONS.get(documentKey);
+        const versionInspections = symbolInspections?.get(symbol.qualifiedName);
+        const symbolVersionId = InspectionCache.calculateSymbolVersionId(document, symbol);
+        return versionInspections?.[0] === symbolVersionId;
+    }
+
+    public static async getCachedInspectionsOfDoc(document: TextDocument): Promise<Inspection[]> {
+        const symbols: SymbolNode[] = await getClassesAndMethodsOfDocument(document);
+        const inspections: Inspection[] = [];
+        for (const symbol of symbols) {
+            const cachedInspections = InspectionCache.getCachedInspectionsOfSymbol(document, symbol);
+            inspections.push(...cachedInspections);
+        }
+        return inspections;
+    }
+
+    /**
+     * @returns the cached inspections, or undefined if not found
+     */
+    public static getCachedInspectionsOfSymbol(document: TextDocument, symbol: SymbolNode): Inspection[] {
+        const documentKey = document.uri.fsPath;
+        const symbolInspections = DOC_SYMBOL_VERSION_INSPECTIONS.get(documentKey);
+        const versionInspections = symbolInspections?.get(symbol.qualifiedName);
+        const symbolVersionId = InspectionCache.calculateSymbolVersionId(document, symbol);
+        if (versionInspections?.[0] === symbolVersionId) {
+            logger.debug(`cache hit for ${SymbolKind[symbol.kind]} ${symbol.qualifiedName} of ${document.uri.fsPath}`);
+            const inspections = versionInspections[1];
+            inspections.forEach(s => {
+                s.document = document;
+                s.problem.position.line = s.problem.position.relativeLine + symbol.range.start.line;
+            });
+            return inspections;
+        }
+        logger.debug(`cache miss for ${SymbolKind[symbol.kind]} ${symbol.qualifiedName} of ${document.uri.fsPath}`);
+        return [];
+    }
+
+    public static cache(document: TextDocument, symbols: SymbolNode[], inspections: Inspection[]): void {
+        for (const symbol of symbols) {
```

Comment:
this updates the cache with the latest symbols in the document, but never has the chance to remove the outdated symbols from the cache. This is not like a sensible cache design to me.

Reply:
changed to **clear cache when user close a document**.
but I don't think it's necessary to clear inspection cache for outdated symbols because it's common case that user make some changes to the code and then ctrl-z to undo them (especially when demo), I keep these inspection cache to avoid repeat inspecting.